# CALIDAD DE DATOS

## IMPORTAR PAQUETES

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from dotenv import load_dotenv
import os
from pathlib import Path
import openpyxl
import sqlalchemy as sa

# Desactivar notación científica
pd.set_option('display.float_format', lambda x: '%.3f' % x)
np.set_printoptions(suppress=True)

# Cargar variables de entorno
load_dotenv()

print("✅ Librerías importadas correctamente")

✅ Librerías importadas correctamente


## IMPORTAR LOS DATOS

In [2]:
df = pd.read_pickle('../02_datos/03_Entrenamiento/train.pkl')
df

,id_cliente,empleo,antigüedad_empleo,ingresos,ingresos_verificados,rating,dti,vivienda,num_hipotecas,num_lineas_credito,...,id_prestamo,descripcion,finalidad,principal,tipo_interes,num_cuotas,imp_cuota,imp_amortizado,estado,imp_recuperado
0,137387967,Hvac technician,3 years,54000.000,Source Verified,A,19.310,MORTGAGE,2.000,10.000,...,NaN,NaN,debt_consolidation,15000.000,7.210,36 months,464.600,2669.060,Current,0.000
1,4798121,"Target Promotions and Marketing,Inc",10+ years,65000.000,Not Verified,D,25.400,RENT,1.000,15.000,...,NaN,NaN,debt_consolidation,10000.000,17.770,36 months,360.380,6362.960,Charged Off,0.000
2,46641215,Banker,5 years,135000.000,Verified,A,14.680,RENT,0.000,19.000,...,NaN,NaN,debt_consolidation,24000.000,6.390,36 months,734.380,24000.000,Fully Paid,0.000
3,87998444,executive director,9 years,188000.000,Source Verified,B,11.690,MORTGAGE,3.000,15.000,...,NaN,NaN,credit_card,27000.000,8.990,60 months,560.350,12443.000,Current,0.000
4,132883631,Subsea Technician,7 years,125000.000,Source Verified,B,9.000,MORTGAGE,1.000,6.000,...,NaN,NaN,debt_consolidation,22000.000,10.900,36 months,719.220,22000.000,Fully Paid,0.000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
199992,73222117,Executive Assistant,10+ years,85000.000,Verified,B,33.050,RENT,1.000,25.000,...,NaN,NaN,debt_consolidation,3000.000,9.750,36 months,96.450,3000.000,Fully Paid,0.000
199995,51876926,Office Manager,10+ years,42000.000,Not Verified,C,20.850,MORTGAGE,6.000,9.000,...,NaN,NaN,debt_consolidation,8000.000,12.290,36 months,266.830,8000.000,Fully Paid,0.000
199996,121031962,Owner & President,6 years,111697.000,Verified,B,16.630,MORTGAGE,2.000,10.000,...,NaN,NaN,other,10000.000,9.440,36 months,320.050,4388.510,Current,0.000
199998,53664762,Attorney,8 years,168000.000,Source Verified,E,4.690,RENT,0.000,8.000,...,NaN,NaN,small_business,30050.000,18.250,60 months,767.170,2964.440,Charged Off,2750.880


In [3]:
# Eliminar el índice del DataFrame
df.reset_index(drop=True, inplace=True)

## VISIÓN GENERAL

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 140000 entries, 0 to 139999
Data columns (total 25 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id_cliente                   140000 non-null  int64  
 1   empleo                       129596 non-null  str    
 2   antigüedad_empleo            130914 non-null  str    
 3   ingresos                     140000 non-null  float64
 4   ingresos_verificados         140000 non-null  str    
 5   rating                       140000 non-null  str    
 6   dti                          139895 non-null  float64
 7   vivienda                     140000 non-null  str    
 8   num_hipotecas                136873 non-null  float64
 9   num_lineas_credito           139998 non-null  float64
 10  porc_tarjetas_75p            135341 non-null  float64
 11  porc_uso_revolving           139893 non-null  float64
 12  num_cancelaciones_12meses    139992 non-null  float64
 13  num_deroga

## TIPOS DE DATOS

### Identificación

In [5]:
df.head().T

,0,1,2,3,4
id_cliente,137387967,4798121,46641215,87998444,132883631
empleo,Hvac technician,"Target Promotions and Marketing,Inc",Banker,executive director,Subsea Technician
antigüedad_empleo,3 years,10+ years,5 years,9 years,7 years
ingresos,54000.000,65000.000,135000.000,188000.000,125000.000
ingresos_verificados,Source Verified,Not Verified,Verified,Source Verified,Source Verified
rating,A,D,A,B,B
dti,19.310,25.400,14.680,11.690,9.000
vivienda,MORTGAGE,RENT,RENT,MORTGAGE,MORTGAGE
num_hipotecas,2.000,1.000,0.000,3.000,1.000
num_lineas_credito,10.000,15.000,19.000,15.000,6.000


In [6]:
df.dtypes

id_cliente                       int64
empleo                             str
antigüedad_empleo                  str
ingresos                       float64
ingresos_verificados               str
rating                             str
dti                            float64
vivienda                           str
num_hipotecas                  float64
num_lineas_credito             float64
porc_tarjetas_75p              float64
porc_uso_revolving             float64
num_cancelaciones_12meses      float64
num_derogatorios               float64
num_meses_desde_ult_retraso    float64
id_prestamo                    float64
descripcion                        str
finalidad                          str
principal                      float64
tipo_interes                   float64
num_cuotas                         str
imp_cuota                      float64
imp_amortizado                 float64
estado                             str
imp_recuperado                 float64
dtype: object

Conclusiones:

* num_hipotecas, num_lineas_credito, num_cancelaciones_12meses, num_derogatorios, num_meses_desde_ult_retraso deberían ser int
* pero vemos que hay nulos, así que no podemos convertirlas hasta que los imputemos

## VALORES ÚNICOS

### Identificación

In [7]:
df.nunique().sort_values()

id_prestamo                         0
num_cuotas                          2
ingresos_verificados                3
vivienda                            6
num_cancelaciones_12meses           7
rating                              7
estado                              9
antigüedad_empleo                  11
finalidad                          14
num_derogatorios                   18
num_hipotecas                      25
num_lineas_credito                 62
num_meses_desde_ult_retraso       131
porc_tarjetas_75p                 170
tipo_interes                      588
porc_uso_revolving               1096
principal                        1472
dti                              5310
descripcion                      7779
imp_recuperado                  10643
ingresos                        11563
imp_cuota                       33575
empleo                          54538
imp_amortizado                  57678
id_cliente                     140000
dtype: int64

Conclusiones:

* id_prestamo con menos de 2 valores --> eliminar

### Corrección

In [8]:
df.drop(columns = 'id_prestamo', inplace = True)

## DUPLICADOS

### Identificación

In [9]:
df.duplicated(subset='id_cliente').sum()

np.int64(0)

## SEPARAR NUMÉRICAS Y CATEGÓRICAS

### Categóricas

In [10]:
cat = df.select_dtypes(exclude = 'number').copy()

### Numéricas

In [11]:
num = df.select_dtypes(include='number').copy()

## GESTIÓN DE CATEGÓRICAS

### Nulos

#### Identificación

In [12]:
cat.isna().sum().sort_values(ascending = False)

descripcion             132199
empleo                   10404
antigüedad_empleo         9086
rating                       0
ingresos_verificados         0
vivienda                     0
finalidad                    0
num_cuotas                   0
estado                       0
dtype: int64

In [13]:
cat.descripcion.value_counts(dropna=False)

descripcion
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     132199
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            15
  Borrower added on 02/25/14 >

In [14]:
cat.empleo.value_counts(dropna=False)

empleo
NaN                             10404
Teacher                          2428
Manager                          2173
Owner                            1376
Registered Nurse                 1000
                                ...  
ILT Inc                             1
Unified Strategies Group            1
TMS Domain Manager                  1
Immigration Senior Paralegal        1
Owner & President                   1
Name: count, Length: 54539, dtype: int64

In [15]:
cat.empleo.nunique()

54538

In [16]:
cat.antigüedad_empleo.value_counts(dropna=False)

antigüedad_empleo
10+ years    46067
2 years      12471
< 1 year     11894
3 years      11260
1 year        9105
NaN           9086
5 years       8674
4 years       8547
6 years       6342
7 years       5849
8 years       5756
9 years       4949
Name: count, dtype: int64

Conclusiones:

* eliminar descripción y empleo
* imputar nulos de antigüedad_empleo por valor 'desconocido'

#### Corrección

In [17]:
# eliminar descripción y empleo
cat.drop(columns = ['descripcion', 'empleo'], inplace = True)

In [18]:
# imputar nulos de antigüedad_empleo por valor 'desconocido'
cat['antigüedad_empleo'] = cat['antigüedad_empleo'].fillna('desconocido')

## GESTIÓN DE NUMÉRICAS

### Nulos

#### Identificación

In [19]:
num.isna().sum().sort_values(ascending = False)

num_meses_desde_ult_retraso    71769
porc_tarjetas_75p               4659
num_hipotecas                   3127
porc_uso_revolving               107
dti                              105
num_cancelaciones_12meses          8
num_lineas_credito                 2
num_derogatorios                   2
ingresos                           0
id_cliente                         0
principal                          0
tipo_interes                       0
imp_cuota                          0
imp_amortizado                     0
imp_recuperado                     0
dtype: int64

Veamoslas con más detalle.

In [20]:
num.num_meses_desde_ult_retraso.value_counts(dropna=False).sort_index()

num_meses_desde_ult_retraso
0.000        185
1.000        439
2.000        626
3.000        800
4.000        965
           ...  
148.000        1
151.000        1
159.000        1
195.000        1
NaN        71769
Name: count, Length: 132, dtype: int64

Hay ceros, pero parecen pocos para una variable que implica retrasos.

Así que vamos a pensar que los nulos significan que no ha habido retrasos.

Y por tanto imputar por ceros.

In [21]:
num.porc_tarjetas_75p.value_counts(dropna=False).sort_index()

porc_tarjetas_75p
0.000      36788
0.170          1
0.330          1
0.500          3
0.570          1
           ...  
94.400         3
94.700         1
95.000         1
100.000    23251
NaN         4659
Name: count, Length: 171, dtype: int64

¿Los 4659 nulos significa que no tienen tarjeta?
La categoría mayoritaria es cero, asi que imput0 por cero.

In [22]:
num.num_hipotecas.value_counts(dropna=False).sort_index()

num_hipotecas
0.000     57547
1.000     24466
2.000     20042
3.000     14404
4.000      9232
5.000      5354
6.000      2911
7.000      1466
8.000       708
9.000       347
10.000      176
11.000       82
12.000       42
13.000       32
14.000       24
15.000       11
16.000        7
17.000        8
18.000        8
19.000        1
20.000        1
21.000        1
25.000        1
27.000        1
32.000        1
NaN        3127
Name: count, dtype: int64

Conclusiones:

* imputar todas por ceros

#### Corrección

In [23]:
num.fillna(0,inplace=True)

,id_cliente,ingresos,dti,num_hipotecas,num_lineas_credito,porc_tarjetas_75p,porc_uso_revolving,num_cancelaciones_12meses,num_derogatorios,num_meses_desde_ult_retraso,principal,tipo_interes,imp_cuota,imp_amortizado,imp_recuperado
0,137387967,54000.000,19.310,2.000,10.000,33.300,45.200,0.000,0.000,10.000,15000.000,7.210,464.600,2669.060,0.000
1,4798121,65000.000,25.400,1.000,15.000,85.700,86.200,0.000,0.000,0.000,10000.000,17.770,360.380,6362.960,0.000
2,46641215,135000.000,14.680,0.000,19.000,0.000,16.300,0.000,0.000,0.000,24000.000,6.390,734.380,24000.000,0.000
3,87998444,188000.000,11.690,3.000,15.000,0.000,3.300,0.000,0.000,0.000,27000.000,8.990,560.350,12443.000,0.000
4,132883631,125000.000,9.000,1.000,6.000,33.300,69.900,0.000,0.000,16.000,22000.000,10.900,719.220,22000.000,0.000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
139995,73222117,85000.000,33.050,1.000,25.000,90.000,76.000,1.000,0.000,0.000,3000.000,9.750,96.450,3000.000,0.000
139996,51876926,42000.000,20.850,6.000,9.000,100.000,81.600,0.000,0.000,13.000,8000.000,12.290,266.830,8000.000,0.000
139997,121031962,111697.000,16.630,2.000,10.000,50.000,26.300,0.000,1.000,6.000,10000.000,9.440,320.050,4388.510,0.000
139998,53664762,168000.000,4.690,0.000,8.000,0.000,52.400,0.000,2.000,0.000,30050.000,18.250,767.170,2964.440,2750.880


Ya habiendo imputado los ceros podemos cambiar los tipos a int, para que ocupen menos memoria.

In [24]:
# meter num_hipotecas, num_lineas_credito, num_cancelaciones_12meses, num_derogatorios, num_meses_desde_ult_retraso en una lista y cambiar su tipo a entero
num_imputar_ceros = ['num_hipotecas', 'num_lineas_credito', 'num_cancelaciones_12meses', 'num_derogatorios', 'num_meses_desde_ult_retraso']
num[num_imputar_ceros] = num[num_imputar_ceros].astype(int)

In [25]:
num.info()

<class 'pandas.DataFrame'>
RangeIndex: 140000 entries, 0 to 139999
Data columns (total 15 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id_cliente                   140000 non-null  int64  
 1   ingresos                     140000 non-null  float64
 2   dti                          140000 non-null  float64
 3   num_hipotecas                140000 non-null  int64  
 4   num_lineas_credito           140000 non-null  int64  
 5   porc_tarjetas_75p            140000 non-null  float64
 6   porc_uso_revolving           140000 non-null  float64
 7   num_cancelaciones_12meses    140000 non-null  int64  
 8   num_derogatorios             140000 non-null  int64  
 9   num_meses_desde_ult_retraso  140000 non-null  int64  
 10  principal                    140000 non-null  float64
 11  tipo_interes                 140000 non-null  float64
 12  imp_cuota                    140000 non-null  float64
 13  imp_amorti

### Atípicos

#### Por desviación típica

##### Identificación

In [26]:
num_desv_tip = 4

Función que devuelve los índices de los atípicos

In [27]:
def atipicos_desv_tip(variable, num_desv_tip = 4):
    #sacamos los nulos por ahora
    variable = variable.dropna()
    #calculamos los límites
    media = np.mean(variable)
    sd = np.std(variable)
    umbral = sd * num_desv_tip
    lim_inf = media - umbral
    lim_sup = media + umbral
    #encontramos los índices de los que están fuera de los límites
    indices = [indice for indice,valor in variable.items() if valor < lim_inf or valor > lim_sup]
    return(indices)

Función que cuenta el número de atípicos

In [28]:
def conteo_atipicos(df,variable, num_desv_tip = 4):
    atipicos = atipicos_desv_tip(df[variable], num_desv_tip)
    return(df.loc[atipicos,variable].value_counts().sort_index())

Variables a aplicar

In [29]:
var_atipicos_dt = ['ingresos','principal','imp_cuota','imp_recuperado']

Obtener los 10 primeros atipicos ordenados en cada variable

In [30]:
for variable in var_atipicos_dt:
    print('\n' + variable + ':\n')
    conteo = conteo_atipicos(num, variable, num_desv_tip)
    print(conteo.sort_index().head(10))


ingresos:

ingresos
415000.000    6
418000.000    1
420000.000    7
422000.000    1
423000.000    2
425000.000    8
426000.000    1
427000.000    1
430000.000    3
431000.000    1
Name: count, dtype: int64

principal:

Series([], Name: count, dtype: int64)

imp_cuota:

imp_cuota
1535.710    1
1566.800    2
1587.020    1
1607.800    1
1714.540    1
Name: count, dtype: int64

imp_recuperado:

imp_recuperado
3227.430    1
3227.610    1
3228.090    1
3228.490    1
3228.600    1
3228.760    1
3229.080    1
3231.130    1
3231.660    1
3231.980    1
Name: count, dtype: int64


Revisamos con más detalle la variable ingresos.

In [31]:
num.ingresos.describe()

count    140000.000
mean      78295.673
std       83346.194
min           0.000
25%       46499.750
50%       65000.000
75%       93600.000
max     9573072.000
Name: ingresos, dtype: float64

##### Corrección

Eliminar los registros con ingresos superiores a 400.000$

*Si se eliminan registros hay que eliminarlos tanto de cat como de num*

###### Id de registros a eliminar

In [32]:
a_eliminar = num.loc[num.ingresos > 400000].index.values
a_eliminar

array([   270,   1238,   1701,   1780,   1816,   3602,   4004,   4145,
         4462,   4569,   4584,   5236,   5258,   5897,   6322,   7163,
         7470,   7719,   8302,   8442,   8598,   8795,   9029,   9165,
         9679,   9847,  10289,  10649,  10865,  11122,  11368,  11454,
        11653,  11917,  12603,  12894,  12975,  13356,  13393,  13505,
        14201,  14790,  15469,  15567,  15629,  16002,  16320,  16364,
        16395,  16457,  16746,  16947,  17016,  17203,  17208,  17480,
        18072,  18564,  18622,  20035,  20355,  20543,  20743,  20902,
        21018,  21086,  21551,  21674,  21944,  22472,  22715,  22751,
        22794,  23551,  23556,  23625,  24139,  24422,  24471,  24546,
        24689,  24793,  25503,  25548,  25687,  25894,  26260,  26334,
        26604,  26783,  26973,  27606,  28387,  28437,  28614,  28954,
        29068,  29247,  29765,  30171,  30205,  30306,  30359,  30994,
        31033,  31478,  31624,  31782,  32622,  33412,  34414,  34877,
      

In [33]:
cat = cat[~cat.index.isin(a_eliminar)]
num = num[~num.index.isin(a_eliminar)]

## GUARDAR DATASETS TRAS CALIDAD DE DATOS

Guarda los avances en df, cat y num con un sufijo para poder guardar o recuperar avances del proyecto.

En formato pickle para no perder las modificaciones de metadatos.

In [34]:
df.to_pickle('../02_datos/03_Entrenamiento/train_calidad.pkl')
cat.to_pickle('../02_datos/03_Entrenamiento/cat_calidad.pkl')
num.to_pickle('../02_datos/03_Entrenamiento/num_calidad.pkl')